cli/bench/cluster.go

```diff
@@ -0,0 +1,231 @@
+	nodesConnectionsPools["replicas"] = make([]RotaryConnectionsPool, len(*ctx.Replicas))
+
+	for i, leader_url := range *ctx.Leaders {
+		connectionsPool, err := createConnectionsPool(ctx, leader_url)
+		if err != nil {
+			return nil, err
+		}
+		nodesConnectionsPools["leaders"][i] = RotaryConnectionsPool{
+			connectionsPool: connectionsPool,
+		}
+	}
+
+	for i, replica_url := range *ctx.Replicas {
+		connectionsPool, err := createConnectionsPool(ctx, replica_url)
+		if err != nil {
+			return nil, err
+		}
+		nodesConnectionsPools["replicas"][i] = RotaryConnectionsPool{
+			connectionsPool: connectionsPool,
+		}
+	}
+
+	return nodesConnectionsPools, nil
+}
+
+// deleteNodesConnectionsPools delete all connections pools.
+func deleteNodesConnectionsPools(nodesConnectionsPools map[string][]RotaryConnectionsPool) {
+	for _, connectionsPools := range nodesConnectionsPools {
+		for i := range connectionsPools {
+			deleteConnectionsPool(connectionsPools[i].connectionsPool)
+		}
+	}
+}
+
+// getNextConnection retrun next connection of node connections pool.
+func (rotaryConnectionsPool *RotaryConnectionsPool) getNextConnection() *tarantool.Connection {
+	rotaryConnectionsPool.mutex.Lock()
+	returnConnection := rotaryConnectionsPool.connectionsPool[rotaryConnectionsPool.currentIndex]
+	rotaryConnectionsPool.currentIndex++
+	rotaryConnectionsPool.currentIndex %= len(rotaryConnectionsPool.connectionsPool)
+	rotaryConnectionsPool.mutex.Unlock()
+	return returnConnection
+}
+
+// getNextConnectionsPool return next node represented by connections pool.
+func (rotaryNodesConnectionsPools *RotaryNodesConnectionsPools) getNextConnectionsPool() *RotaryConnectionsPool {
+	rotaryNodesConnectionsPools.mutex.Lock()
+	returnConnectionsPool := &rotaryNodesConnectionsPools.rotaryConnectionsPool[rotaryNodesConnectionsPools.currentIndex]
+	rotaryNodesConnectionsPools.currentIndex++
+	rotaryNodesConnectionsPools.currentIndex %= len(rotaryNodesConnectionsPools.rotaryConnectionsPool)
+	rotaryNodesConnectionsPools.mutex.Unlock()
+	return returnConnectionsPool
+}
+
+// benchCluster execute bench algorithm for cluster.
+func benchCluster(ctx context.BenchCtx, benchData *BenchmarkData) error {
+	// Сreate connections pools for all nodes in cluster before starting the benchmark
+	// to exclude the connection establishment time from measurements.
+	nodesConnectionsPools, err := createNodesConnectionsPools(ctx)
+	if err != nil {
+		return err
+	}
+	defer deleteNodesConnectionsPools(nodesConnectionsPools)
+
+	mutationConnections := nodesConnectionsPools["leaders"]
+	selectConnections := nodesConnectionsPools["replicas"]
+
+	if len(*ctx.Replicas) == 0 {
+		selectConnections = append(selectConnections, mutationConnections...)
+	}
+
+	benchData.startTime = time.Now()
+
+	// Start detached connections.
+	for i := 0; i < ctx.Connections; i++ {
+		benchData.waitGroup.Add(1)
+		go func() {
+			defer benchData.waitGroup.Done()
+			requestsSequence := RequestsSequence{
```

Comment:
Now requests generation is the part of the bench run tracked by timer. I think it's better to build request data before starting the timer, it possible.